File: app/src/main/java/com/bortolan/iquadriv2/Utils/DownloadSchedules.java
package com.bortolan.iquadriv2.Utils;


import android.os.AsyncTask;

import com.bortolan.iquadriv2.Interfaces.GitHub.GitHubResponse;
import com.google.gson.Gson;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.URL;


public class DownloadSchedules extends AsyncTask<Void, Void, GitHubResponse> {
    private Execute execute;

    public DownloadSchedules(Execute execute) {
        this.execute = execute;
    }

    @Override
    protected GitHubResponse doInBackground(Void... voids) {
        GitHubResponse response = null;
        try {
            InputStream inputStream = new URL("https://raw.githubusercontent.com/Quix0tic/iQuadri/master/orari.json").openConnection().getInputStream();

            if (inputStream != null) {
                BufferedReader r = new BufferedReader(new InputStreamReader(inputStream));
                StringBuilder total = new StringBuilder();
                String line;
                while ((line = r.readLine()) != null) {
                    total.append(line).append('\n');
                }
                response = new Gson().getAdapter(GitHubResponse.class).fromJson(total.toString());
            }
        } catch (Exception e) {
            e.printStackTrace();
        }

        return response;
    }

    @Override
    protected void onPostExecute(GitHubResponse response) {
        super.onPostExecute(response);
        execute.onPost(response);
    }

    public interface Execute {
        void onPost(GitHubResponse response);
    }
}

File: app/src/main/java/com/bortolan/iquadriv2/Tasks/Remote/DownloadSchedules.kt
package com.bortolan.iquadriv2.Tasks.Remote


import android.os.AsyncTask
import com.bortolan.iquadriv2.Interfaces.GitHub.GitHubResponse
import com.google.gson.Gson
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.URL


class DownloadSchedules(private val execute: (GitHubResponse) -> Unit) : AsyncTask<Void, Void, GitHubResponse?>() {

    override fun doInBackground(vararg voids: Void): GitHubResponse? {
        var response: GitHubResponse? = null
        try {
            val inputStream = URL(SCHEDULES).openConnection().getInputStream()

            if (inputStream != null) {
                val r = BufferedReader(InputStreamReader(inputStream))
                val total = StringBuilder()
                var line: String?
                while (true) {
                    line = r.readLine()
                    if (line == null) break
                    total.append(line).append('\n')
                }
                response = Gson().getAdapter(GitHubResponse::class.java).fromJson(total.toString())
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }

        return response
    }

    override fun onPostExecute(response: GitHubResponse?) {
        super.onPostExecute(response)
        if (response != null) execute.invoke(response)
    }

    companion object {
        val SCHEDULES = "https://liceo.quadri.bortolan.ml"
    }
}
